documentation/MSBuild-scheduler.md

```diff
@@ -0,0 +1,171 @@
+# MSBuild Scheduler Architecture
+
+The diagram below illustrates the key components and workflows of the MSBuild Scheduler system.
+
+![MSBuild Scheduler Schema](./scheduler_architecture.png)
+
+## Request Lifecycle
+
+### 1. Submission
+- BuildManager generates root request to the Scheduler
+- Configuration is registered and SchedulableRequest is created
+
+### 2. Scheduling
+- Scheduler applies algorithm to select node for request
+- Request is assigned to node and moves to Executing state
+
+### 3. Execution
+- Node processes request and may encounter dependencies
+- If blocked, node reports to Scheduler with blocker information
+- Scheduler creates new requests for dependencies and schedules them
+
+### 4. Completion
+- Node reports build results back to Scheduler
+- Results are cached and request is marked complete
+- Parent requests are unblocked with results
+
+### 5. Result Processing
+- Results propagate through dependency chain
+- When all requests complete, submission is reported complete
+
+## Core Components
+
+### SchedulableRequest
+The SchedulableRequest is a crucial internal wrapper class in the MSBuild scheduler that manages the lifecycle of build requests.
+It tracks:
+
+- The current state of the request (Unscheduled, Executing, Blocked, Ready, Yielding, Completed)
+- Which node it's assigned to
+- The relationships to other requests (parent/child or blocking/blocked-by)
+- Timing and execution details
+
+Think of `BuildRequest` as the "what" and `SchedulableRequest` as the "how" to build a project.
+
+#### When SchedulableRequest has a parent?
+In MSBuild, builds often have dependencies. When a project needs to build another project first, it creates a child request.
+The original one becomes the "parent request" of this new child request.
+For example:
+
+`ProjectA` is building (parent request) => `ProjectA` references `ProjectB`, so it creates a request to build `ProjectB` (child request) => `ProjectA's` request is considered the "parent request" of `ProjectB's` request
+
+This parent-child relationship is tracked through:
+- `ParentGlobalRequestId` in the `BuildRequest`
+- The `Parent` property in the `SchedulableRequest`
+
+The parent request is blocked while its child request is executing. This is a fundamental aspect of MSBuild's dependency handling system.
+When a parent request determines it needs results from a child request, the following happens:
+
+1) The parent request enters the "Blocked" state
+2) The scheduler creates the child request
+3) The parent remains blocked until the child completes
+4) Once the child request finishes, its results are returned to the parent
+5) The parent moves from "Blocked" to "Ready" state
+6) The scheduler can then resume executing the parent request
+
+![MSBuild Scheduler Schema](./schedulableRequest_parent_relationships.png)
+
+If the parent is absent, meaning a schedulable request doesn't have any dependencies, the request will have `parentRequest = null`.
```

Comment:
"If the parent is absent, meaning a schedulable request doesn't have any dependencies" - this does not seem right. If the project has no dependencies, the request will not have children, not parents.